cpp/src/arrow/util/spaced.h

```diff
@@ -0,0 +1,199 @@
+    }
+    idx_src++;
+    valid_bits_offset++;
+  }
+
+  // The aligned parts scaned with BitBlockCounter
+  arrow::internal::BitBlockCounter data_counter(valid_bits, valid_bits_offset,
+                                                num_values - leading_bits);
+  auto current_block = data_counter.NextWord();
+  while (idx_src < num_values) {
+    if (current_block.AllSet()) {  // All true values
+      int run_length = 0;
+      // Scan forward until a block that has some false values (or the end)
+      while (current_block.length > 0 && current_block.AllSet()) {
+        run_length += current_block.length;
+        current_block = data_counter.NextWord();
+      }
+      // Fill all valid values of this scan
+      std::memcpy(&output[num_valid_values], &src[idx_src], run_length * sizeof(T));
+      num_valid_values += run_length;
+      idx_src += run_length;
+      valid_bits_offset += run_length;
+      // The current_block already computed, advance to next loop
+      continue;
+    } else if (!current_block.NoneSet()) {  // Some values are null
+      arrow::internal::BitmapReader valid_bits_reader(valid_bits, valid_bits_offset,
+                                                      current_block.length);
+      for (int64_t i = 0; i < current_block.length; i++) {
+        if (valid_bits_reader.IsSet()) {
+          output[num_valid_values] = src[idx_src];
+          num_valid_values++;
+        }
+        idx_src++;
+        valid_bits_reader.Next();
+      }
+      valid_bits_offset += current_block.length;
+    } else {  // All null values
+      idx_src += current_block.length;
+      valid_bits_offset += current_block.length;
+    }
+    current_block = data_counter.NextWord();
+  }
+
+  return num_valid_values;
+}
+
+/// \brief Expand the buffer but leave spaces for null entries.
+///
+/// \param[in, out] buffer the in-place buffer
+/// \param[in] num_values total size of buffer including null slots
+/// \param[in] null_count number of null slots
+/// \param[in] valid_bits bitmap data indicating position of valid slots
+/// \param[in] valid_bits_offset offset into valid_bits
+/// \return The number of values expanded, including nulls.
+template <typename T>
+inline int SpacedExpand(T* buffer, int num_values, int null_count,
+                        const uint8_t* valid_bits, int64_t valid_bits_offset) {
+  constexpr uint64_t kBatchSize = 64;
+  static_assert(kBatchSize == 64, "Invalid batch size, BitBlockCounter stick to 64");
+  constexpr uint64_t kBatchByteSize = kBatchSize / 8;
+
+  // Point to end as we add the spacing from the back.
+  int idx_decode = num_values - null_count - 1;
+  int idx_buffer = num_values - 1;
+  int64_t idx_valid_bits = valid_bits_offset + idx_buffer;
+
+  // Depending on the number of nulls, some of the value slots in buffer may
+  // be uninitialized, and this will cause valgrind warnings / potentially UB
+  std::memset(static_cast<void*>(&buffer[idx_decode + 1]), 0, null_count * sizeof(T));
```

Comment:
is this `static_cast` needed?

Reply:
/arrow/cpp/src/arrow/util/spaced.h:119:14: error: 'void* memset(void*, int, size_t)' clearing an object of non-trivial type 'struct parquet::ByteArray'; use assignment or value-initialization instead [-Werror=class-memaccess]

It get a error without the cast on "C GLib & Ruby / AMD64 Ubuntu 20.04 GLib & Ruby" job

Reply:
Ah that's right. Thanks